pkg/tenant/topics.go

```diff
@@ -10,29 +10,68 @@ import (
 	"github.com/rs/zerolog/log"
 )
 
+// ProjectTopicList retrieves all topics assigned to a project
+// and returns a 200 OK response.
+//
+// Route: /projects/:projectID/topics
 func (s *Server) ProjectTopicList(c *gin.Context) {
-	c.JSON(http.StatusNotImplemented, "not implemented yet")
+	var (
+		err error
+	)
+
+	// Get the topic's project ID from the URL and return a 400 response
+	// if the project ID is not a ULID.
+	var projectID ulid.ULID
+	if projectID, err = ulid.Parse(c.Param("projectID")); err != nil {
+		log.Error().Err(err).Msg("could not parse project ulid")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("could not parse project ulid"))
+		return
+	}
+
+	// Get topics from the database and return a 500 response
+	// if not successful.
+	if _, err = db.ListTopics(c.Request.Context(), projectID); err != nil {
+		log.Error().Err(err).Msg("could not fetch topics from the database")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not fetch topics from the database"))
+		return
+	}
+
+	// Build the response.
+	out := &api.ProjectTopicPage{TenantTopics: make([]*api.Topic, 0)}
+
+	projectTopic := &api.Topic{}
+
+	out.TenantTopics = append(out.TenantTopics, projectTopic)
+
+	c.JSON(http.StatusOK, out)
 }
 
 func (s *Server) ProjectTopicCreate(c *gin.Context) {
 	c.JSON(http.StatusNotImplemented, "not implemented yet")
 }
 
+// TopicList retrieves all topics assigned to a project and
+// returns a 200 OK response.
+//
+// Route: /topics
 func (s *Server) TopicList(c *gin.Context) {
-	// The following TODO task items will need to be
-	// implemented for each endpoint.
+	// TODO: Fetch the topic's project ID from key.
 
-	// TODO: Add authentication and authorization middleware
-	// TODO: Identify top-level info
-	// TODO: Parse and validate user input
-	// TODO: Perform work on the request, e.g. database interactions,
-	// sending notifications, accessing other services, etc.
+	// Get topics from the database and return a 500 response
+	// if not successful.
+	if _, err := db.ListTopics(c.Request.Context(), ulid.ULID{}); err != nil {
```

Comment:
Looks like we will want to store the topics (`[]*Topic`) passed back by the database here so that we can assign the result to an `out` attribute

Reply:
Also, perhaps we should create the `ulid.ULID{}` right after line 58 so that we remember that this should be a value parsed from the key rather than an empty struct in the future?